pkg/sentry/fsimpl/fuse/connection.go

```diff
@@ -84,16 +205,38 @@ func NewFUSEConnection(_ context.Context, fd *vfs.FileDescription, maxInFlightRe
 	fuseFD.fullQueueCh = make(chan struct{}, maxInFlightRequests)
 	fuseFD.writeCursor = 0
 
-	return &Connection{
-		fd: fuseFD,
+	return &connection{
+		fd:                  fuseFD,
+		maxBackground:       fuseDefaultMaxBackground,
+		congestionThreshold: fuseDefaultCongestionThreshold,
+		maxPages:            fuseDefaultMaxPagesPerReq,
+		initializedChan:     make(chan struct{}),
+		connected:           true,
 	}, nil
 }
 
+// SetInitialized atomically sets the connection as initialized.
+func (conn *connection) SetInitialized() {
+	// Unblock the requests sent before INIT.
+	// Close the channel first to ensure the
+	// tasks that raced before the initialized is
+	// set will also not block.
+	close(conn.initializedChan)
+
+	atomic.StoreInt32(&(conn.initialized), int32(1))
```

Comment:
Should we not do this before closing the channel? What I'm worried about is that "bargers" or new requests will try and Block on a closed channel after seeing that the fs is not initialized. That will error. Instead we should close the channel after we mark it as initialized so new requests are not aborted (if the happen in between the last two lines).

Reply:
Exactly contrary, those requests saw a non-initialized will see a channel that has been closed and thus no longer blocking.
```golang
	c := make(chan struct{})
	close(c)
	select {
	case <-c:
		fmt.Println("will print this")
	default:
	}
```

This has the added benefit of maintaining the orders of the requests

Reply:
I don't think that this will maintaining the orders of the requests. You can try to run this:

```
	c := make(chan int, 1)
	for i := 0; i < 100; i++ {
		go func(i int) {
			<-c
			fmt.Printf("%d\n", i)
		}(i)
	}
	close(c)
	time.Sleep(time.Second)
```

I agree with @ridwanmsharif  that closing the channel after setting initialized looks more logical. If you think your way is better, you need to write a comment in the code which explains why it works this way.

Reply:
The order I'm trying to maintain is the order between the group of blocked requests and the group of requests that are issued right when it's unblocking. The first group should be sent earlier but if the initialize bool is set first, they will have a higher probability to be issued after the second group